lightning/src/util/ser_macros.rs

```diff
@@ -7,82 +7,142 @@
 // You may not use this file except in accordance with one or both of these
 // licenses.
 
+macro_rules! encode_a_tlv {
+	($stream: expr, $type: expr, $field: expr, explicit) => {
+		BigSize($type).write($stream)?;
+		BigSize($field.serialized_length() as u64).write($stream)?;
+		$field.write($stream)?;
+	};
+	($stream: expr, $type: expr, $field: expr, vec_type) => {
+		encode_a_tlv!($stream, $type, ::util::ser::VecWriteWrapper(&$field), explicit);
+	};
+	($stream: expr, $optional_type: expr, $optional_field: expr, option) => {
+		if let Some(ref field) = $optional_field {
+			BigSize($optional_type).write($stream)?;
+			BigSize(field.serialized_length() as u64).write($stream)?;
+			field.write($stream)?;
+		}
+	};
+}
+
 macro_rules! encode_tlv {
-	($stream: expr, {$(($type: expr, $field: expr)),*}, {$(($optional_type: expr, $optional_field: expr)),*}) => { {
+	($stream: expr, {$(($type: expr, $field: expr, $fieldty: ident)),*}) => { {
 		#[allow(unused_imports)]
 		use util::ser::BigSize;
-		// Fields must be serialized in order, so we have to potentially switch between optional
-		// fields and normal fields while serializing. Thus, we end up having to loop over the type
-		// counts.
-		// Sadly, while LLVM does appear smart enough to make `max_field` a constant, it appears to
-		// refuse to unroll the loop. If we have enough entries that this is slow we can revisit
-		// this design in the future.
-		#[allow(unused_mut)]
-		let mut max_field: u64 = 0;
-		$(
-			if $type >= max_field { max_field = $type + 1; }
-		)*
 		$(
-			if $optional_type >= max_field { max_field = $optional_type + 1; }
+			encode_a_tlv!($stream, $type, $field, $fieldty);
```

Comment:
Can you add an assertion that the fields are in type-order? It shouldn't be too hard to add a mutable last_field variable outside the repetition and check+update it inside, I don't think rustc will detect the issue at compile-time but at least it will be guaranteed to fail at runtime (maybe `debug_assert`?).